contributors/design-proposals/no-new-privs.md

```diff
@@ -0,0 +1,107 @@
+
+### Interactions with other Linux primitives
+
+- suid binaries: will break when `no_new_privs` is enabled
+- seccomp2 as a non root user: requires `no_new_privs`
+- seccomp2 with dropped `CAP_SYS_ADMIN`: requires `no_new_privs`
+- ambient capabilities: requires `no_new_privs`
+- selinux transactions: bugs that were fixed documented [here](https://github.com/moby/moby/issues/23981#issuecomment-233121969)
+
+
+## Current Implementations
+
+### Support in Docker
+
+Since Docker 1.11, a user can specify `--security-opt` to enable `no_new_privs`
+while creating containers, for example
+`docker run --security-opt=no_new_privs busybox`.
+
+Docker provides via their Go api an object named `ContainerCreateConfig` to
+configure container creation  parameters. In this object, there is a string
+array `HostConfig.SecurityOpt` to specify the security options. Client can
+utilize this field to specify the arguments for security options while
+creating new containers.
+
+This field did not scale well for the Docker client, so it's suggested that
+Kubernetes does not follow that design.
+
+More details of the Docker implementation can be read
+[here](https://github.com/moby/moby/pull/20727) as well as the original
+discussion [here](https://github.com/moby/moby/issues/20329).
+
+### Support in rkt
+
+Since rkt v1.26.0, the `NoNewPrivileges` option has been enabled in rkt.
+
+More details of the rkt implementation can be read
+[here](https://github.com/rkt/rkt/pull/2677).
+
+### Support in OCI runtimes
+
+Since version 0.3.0 of the OCI runtime specification, a user can specify the
+`noNewPrivs` boolean flag in the configuration file.
+
+More details of the OCI implementation can be read
+[here](https://github.com/opencontainers/runtime-spec/pull/290).
+
+## Existing SecurityContext objects
+
+Kubernetes defines `SecurityContext` for `Container` and `PodSecurityContext`
+for `PodSpec`. `SecurityContext` objects define the related security options
+for Kubernetes containers, e.g. selinux options.
+
+To support "no new privileges" options in Kubernetes, it is proposed to make
+the following changes:
+
+## Changes of SecurityContext objects
+
+Add a new bool type field named `allowPrivilegeEscalation` to both `SecurityContext`
+definition and `PodSecurityContext` definition:
+
+* For example, `allowPrivilegeEscalation=true` in `PodSecurityContext`
+means that all the containers in the pod should **NOT** be run with
+`no_new_privs` enabled. This enables pod level control of `no_new_privs` flag.
+* `allowPrivilegeEscalation` in `SecurityContext` provides container level
+control of the `no_new_privs` flag and can override the pod level
+`allowPrivilegeEscalation` setting.
+
+By default, `allowPrivilegeEscalation` is `false` at the kubelet level
+with the following exceptions:
+
+- when a container is `privileged`
+- when `CAP_SYS_ADMIN` is added to a container
+- when a container is not run as root, uid `0` (to prevent breaking suid
```

Comment:
I think we are agreeing on the fundamentals. 

The `securityContext` field is the override.  If it has a value then the kubelet honors it.  If it is nil then the kubelet has defaulting logic.

PSP controls whether or not you can set the value in `securityContext` and thereby if you can trigger the kubelet logic.    

When `PSP.allowPrivilegeEscalation=true` you can set it to true, false, or leave it nil.  You're free to override the setting or get the defaulting behavior.  There is no PSP defaulting in this scenario.

When `PSP.allowPrivilegeEscalation=false` you cannot set it to true (deny in admission) and if you leave it nil then PSP will default it to `false` to make sure the kubelet defaulting isn't triggered.

That allows admins to prevent any scenario where they want to ensure escalation never happens and also allow scenarios where users can do what they please or allow the system to DTRT.